gax/src/query_parameter.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub type Error = serde_json::Error;
+type Result = serde_json::Result<String>;
+
+pub fn format<T>(
+    name: &'static str,
+    parameter: &T,
+) -> serde_json::Result<Option<(&'static str, String)>>
+where
+    T: QueryParameter,
+{
+    QueryParameter::format(parameter)
+        .map(|result| result.map(|s| (name, s)))
+        .transpose()
+}
+
+pub trait QueryParameter {
+    fn format(&self) -> Option<Result>;
+}
+
+impl<T: QueryParameter> QueryParameter for Option<T> {
+    fn format(&self) -> Option<Result> {
+        self.as_ref().and_then(|v| QueryParameter::format(v))
+    }
+}
+
+impl<T: RequiredQueryParameter> QueryParameter for T {
+    fn format(&self) -> Option<Result> {
+        Some(RequiredQueryParameter::format(self))
+    }
+}
+
+/// Format query parameters as strings.
+trait RequiredQueryParameter {
+    fn format(&self) -> Result;
+}
+
+impl RequiredQueryParameter for String {
+    fn format(&self) -> Result {
+        Ok(self.clone())
+    }
+}
+
+impl RequiredQueryParameter for i32 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for u32 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for i64 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for u64 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for f32 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for f64 {
+    fn format(&self) -> Result {
+        Ok(format!("{self}"))
+    }
+}
+
+impl RequiredQueryParameter for types::Duration {
+    fn format(&self) -> Result {
+        Ok(serde_json::to_value(self)?.as_str().unwrap().to_string())
```

Comment:
Since we are already calling unwrap here and we control the input type, could we simplify this and remove the need to return a result at all? I think we can guarantee that  our types will always serialize.

Reply:
That `unwrap` is just turning `Option<&str>` into `&str`.  I am using secret knowledge about how `Duration` always serializes to a JSON string.  Maybe a cleaner approach would be to provide a function in `Duration` to format as a string, and then use that here (and in the serializer).

To your larger point: I am not sure it is true that we can always serialize the values we have. For example:  Timestamps can only be (safely) serialized in a certain range:

https://github.com/protocolbuffers/protobuf/blob/d0aefcba0ea1dd814c12915685eda9e953df49bb/src/google/protobuf/timestamp.proto#L53-L55

We currently do not enforce this restriction, but I can see us returning an error if the timestamp cannot be formatted correctly because it is out of range.

Reply:
hmm that is fair, proceed!